pageserver/src/remote_storage/storage_sync/download.rs

```diff
@@ -43,53 +43,67 @@ pub(super) async fn download_timeline<
-            retries,
-            SyncKind::Download(download),
-        ));
-        return Some(false);
-    }
 
     let TimelineSyncId(tenant_id, timeline_id) = sync_id;
-
     let index_read = remote_assets.1.read().await;
     let remote_timeline = match index_read.timeline_entry(&sync_id) {
         None => {
             error!("Cannot download: no timeline is present in the index for given ids");
             return None;
         }
-        Some(TimelineIndexEntry::Full(remote_timeline)) => Cow::Borrowed(remote_timeline),
-        Some(TimelineIndexEntry::Description(_)) => {
-            drop(index_read);
-            debug!("Found timeline description for the given ids, downloading the full index");
-            match update_index_description(
-                remote_assets.as_ref(),
-                &conf.timeline_path(&timeline_id, &tenant_id),
-                sync_id,
-            )
-            .await
-            {
-                Ok(remote_timeline) => Cow::Owned(remote_timeline),
-                Err(e) => {
-                    error!("Failed to download full timeline index: {:#}", e);
-                    sync_queue::push(SyncTask::new(
+        Some(index_entry) => {
+            let remote_disk_consistent_lsn = index_entry.disk_consistent_lsn()?;
+            match index_entry {
+                TimelineIndexEntry::Full(remote_timeline) => Cow::Borrowed(remote_timeline),
+                TimelineIndexEntry::Description(_) => {
+                    drop(index_read);
+                    debug!(
+                        "Found timeline description for the given ids, downloading the full index"
+                    );
+                    match update_index_description(
+                        remote_assets.as_ref(),
+                        &conf.timeline_path(&timeline_id, &tenant_id),
                         sync_id,
-                        retries,
-                        SyncKind::Download(download),
-                    ));
-                    return Some(false);
+                    )
+                    .await
+                    {
+                        Ok(remote_timeline) => Cow::Owned(remote_timeline),
+                        Err(e) => {
+                            error!("Failed to download full timeline index: {:#}", e);
+                            sync_queue::push(SyncTask::new(
+                                sync_id,
+                                retries,
+                                SyncKind::Download(download),
+                            ));
+                            return Some((false, remote_disk_consistent_lsn));
+                        }
+                    }
                 }
             }
         }
     };
+    let remote_disk_consistent_lsn = match remote_timeline.checkpoints().max() {
+        Some(lsn) => lsn,
+        None => {
+            debug!("Cannot download: no disk consistent Lsn is present for the remote timeline");
+            return None;
```

Comment:
I'm also not sure how applicable those cases are, I'm using `DownloadedTimeline::Abort` instead of `None` to highlight those cases after your proposals (check it out), and there are 3 cases where I use it, all related to situations when the index data about the remote timelines is incorrect: either misses the remote timeline id, or the entire timeline for the id given.

Both at this point seem more like a programmatic error rather than a possible case, but I'm super reluctant to panicking in such cases, preferring to return some error instead, that could be logged and processed regularly.

Reply:
Thanks! I like the explicitness of DownloadedTimeline::Abort

Agree that these are more of a programmatic errors and Result is preferable over panics